File: HW3/cppstack/stack.cpp
#include "stack.h"
#include <memory>
#include <string>

StringResponse::StringResponse(const std::string& value, ResponseCode code)
  : value(value), code(code) {}

StringResponse::~StringResponse() {}

Stack::Stack() : stack(std::make_unique<std::string[]>(STARTING_STACK_SIZE)), top(0), array_length(STARTING_STACK_SIZE) {}

Stack::~Stack() {}

ResponseCode Stack::push(const std::string& value) {
  if (top == MAX_STACK_SIZE) {
    return ResponseCode::StackFull;
  }

  if (isFull()) {
    // array_length += STARTING_STACK_SIZE;
    array_length = array_length * 2;
    if (array_length > MAX_STACK_SIZE) {
      return ResponseCode::OutOfMemory;
    }
    std::unique_ptr<std::string[]> resized_stack = std::make_unique<std::string[]>(array_length);
    for (size_t i = 0; i < top; i++) {
      resized_stack[i] = stack[i];
    }

    stack.swap(resized_stack);
  }
  stack[top++] = value;
  return ResponseCode::Success;
}

StringResponse Stack::peek() const {
  if (isEmpty()) {
    return StringResponse("", ResponseCode::StackEmpty);
  }
  return StringResponse(stack[top - 1], ResponseCode::Success);
}

StringResponse Stack::pop() {
  if (isEmpty()) {
    return StringResponse("", ResponseCode::StackEmpty);
  }
  
  std::string value = stack[--top];

  if (top > STARTING_STACK_SIZE && top == array_length / 4) {
    array_length /= 2;
    std::unique_ptr<std::string[]> resized_stack = std::make_unique<std::string[]>(array_length);
    for (size_t i = 0; i < top; i++) {
      resized_stack[i] = stack[i];
    }
    stack.swap(resized_stack);
  }

  return StringResponse(value, ResponseCode::Success);
}

bool Stack::isEmpty() const {
  return top == 0;
}

bool Stack::isFull() const {
  return top == array_length;
}

size_t Stack::size() const {
  return top;
}

File: HW3/jstack/java_stack/src/main/java/jstack/Stack.java
public class Stack {
    private static final int MAX_CAPACITY = 65536;
    private static final int STARTING_STACK_SIZE = 128;
    private String[] elements;
    private int size;

    public Stack() {
        elements = new String[STARTING_STACK_SIZE];
        size = 0;
    }

    public boolean isFull() {
        return size == elements.length;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public int size() {
        return size;
    }

    public void push(String item) {
        if (size == MAX_CAPACITY) {
            throw new IllegalStateException("Stack is full");
        } else if (isFull()) {
            final String[] resizedStack = new String[elements.length * 2];
            for (int i = 0; i < elements.length; i++) {
                resizedStack[i] = elements[i];
            }
            elements = resizedStack;
        }
        elements[size++] = item;
    }

    public String pop() {
        if (isEmpty()) {
            throw new IllegalStateException("Stack is empty");
        }
        return elements[--size];
    }

    public String peek() {
        if (isEmpty()) {
            throw new IllegalStateException("Stack is empty");
        }
        return elements[size - 1];
    }
}
